python/ray/data/datasource/datasink.py

```diff
@@ -67,44 +46,32 @@ def write(
         self,
         blocks: Iterable[Block],
         ctx: TaskContext,
-    ) -> None:
+    ) -> WriteResultType:
```

Comment:
@bveeramani Unrelated to this PR. But one pitfall about the current Datasink interface is that, the Datasink object will be used both on the driver (the `on_xxx` callbacks) and on the workers (this `write` function). 
Users may mistakenly think that if they update an attribute in the write method, the update will be available on `on_write_complete`. 

We should consider addressing this issue before making the Datasink API public. One solution is to introduce a separate `DatasinkWriter` class.

Reply:
Yeah, I agree, that's janky.

>  One solution is to introduce a separate DatasinkWriter class.

Sounds reasonable.

> We should consider addressing this issue before making the Datasink API public. 

Makes sense. There's no urgency to make `Datasink` public.